kernel/src/ipc/semary.rs

```diff
@@ -0,0 +1,77 @@
+use crate::sync::Semaphore;
+use crate::sync::SpinLock as Mutex;
+use alloc::{boxed::Box, collections::BTreeMap, string::String, sync::Arc, sync::Weak, vec::Vec};
+use core::cell::UnsafeCell;
+use lazy_static::lazy_static;
+use spin::RwLock;
+
+pub trait SemArrTrait {
+    fn get_x(&self, x: usize) -> &Semaphore;
+}
+
+pub struct SemArray {
+    pub key: usize,
+    pub sems: Vec<Semaphore>,
+}
+
+unsafe impl Sync for SemArray {}
+unsafe impl Send for SemArray {}
+
+impl SemArray {
+    pub fn new(key: usize, sems: Vec<Semaphore>) -> SemArray {
+        SemArray {
+            key: key,
+            sems: sems,
+        }
+    }
+}
+
+impl SemArrTrait for SemArray {
+    fn get_x(&self, x: usize) -> &Semaphore {
+        &self.sems[x]
+    }
+}
+
+pub struct SemBuf {
```

Comment:
Should add an annotation `#[repr(C)]` because it's passed from syscall argument.